### ISSUE TITLE
Losing Lambda invocation traces with goroutines.

### ISSUE THREAD

**Issue by daroot** [aws/aws-xray-sdk-go#87]
I'm not 100% sure this a bug, or me failing to correctly intuit some interaction between Go contexts, goroutines, and aws-xray-sdk-go's interaction with them for storing trace data.

I have a lambda that is invoking other lambdas as a warmer function.  Lambda A (the originating lambda) calls other lambdas (B, C, D, E, etc).

This linear and synchronous code works perfectly, showing me every trace I expect to see, with A's timeline reporting sequential calls to the Lambda service for invoking, B, C, D, E one after another in the timeline:

```go
l := lambda.New(session.Must(session.NewSession()))
xray.AWS(l.Client)
ctx := context.Background()
for _, c := range funcConfigs {
    for x := 0; x < 8; x++ {
         l.InvokeWithContext(ctx, &lambda.InvokeInput{
             FunctionName: &c.Name,
             InvocationType: aws.String(lambda.InvocationTypeEvent),
             Payload: []byte(c.Payload),
    }
}
```
The exact same code, but running in parallel with goroutines does not:
```go
l := lambda.New(session.Must(session.NewSession()))
xray.AWS(l.Client)
ctx := context.Background()
wg := &sync.WaitGroup{}
for _, c := range funcConfigs {
    c := c
    for x := 0; x < 8; x++ {
        wg.Add(1)
        go func() {
             l.InvokeWithContext(ctx, &lambda.InvokeInput{
                 FunctionName: &c.Name,
                 InvocationType: aws.String(lambda.InvocationTypeEvent),
                Payload: []byte(c.Payload),
             }
            wg.Done()
        }()
    }
}
wg.Wait()
```

Instead, what I see is that A's timeline reports usually one or sometimes two entries to Lambda, for a random member of the B/C/D/E subset, despite there being 4 separate successful calls to lambda.InvokeWithContext.   Likewise, the service map shows disconnected nodes calls from 'Clients' for the resulting calls in whichever of B, C, D, and E was not reported in A's timeline.

The target lambda invocations _do_ show up in the overall trace with the correct trace_id, including the AWS::Lambda Dwell/Attempts and their own AWS::Lambda::Function timeline, but the calls from A to Lambda (with the breakdown of marshal/attempt/connect/unmarshal/etc) that started each are missing from A's timeline.

Any time I see behavior where it looks like something's getting overwritten in a for loop that launches goroutines, I'm immediately suspicious of [closure and variable shadowing](https://golang.org/doc/faq#closures_and_goroutines).   Adding explicit copies of the context inside the inner for loop did not help.  Assuming this is because a copy of a context.Context interface is just copying pointers, I tried context.WithValue to produce a child context to pass to lamba.InvokeWithContext, and that did not change the behavior either.   using xray.CaptureAsync around the InvokeWithContext call did cause the generated subsegments to appear, but they did not end up as parents to the calls to InvokeWithContext, and I did not get the expect 5 total copies of marshal/attempt/etc.

xray.DetachContext doesn't seem to do the job (and in fact, prevents the downstream traces from even showing up in originating lambda's overall trace data).

What's the magic incarnation I need in order for the originating lambda to show multiple parallel lambda.InvokeWithContext calls to each of B, C, D, and E in its timeline?

I've attached a trace showing the failure mode, with the appropriate lambda names, dynamo tables, and account-ids scrubbed; the trace-ids are real, if you'd like to investigate directly.  [trace.txt](https://github.com/aws/aws-xray-sdk-go/files/2939146/trace.txt)

_edit: make the number of downstream lambdas match the included trace_

**Comment by luluzhao** [comment]
Hello @daroot , sorry for the late reply. After reading this issue, it seems like there are several unexpected behaviors here. Please correct if I am wrong. 1. subsegments B, C, D, E were missing from A's timeline. 2. the breakdown of marshal/attempt/connect/unmarshal/etc were also missing from the A's timeline. Do I miss something else?

Would you mind providing the screenshot of the two cases to me since raw trace doc is too long which is hard to understand this cause and diagnose. If not, may I use the code snippet you attached above for diagnosing this issue? Does the code snippet cover this issue?

**Comment by daroot** [comment]
There's basically one bad behavior; when using goroutines, the  warmer function (A) is not showing the correct number of subtrees for the invocation of the downstream Lambdas (B, C, D, E) in the timeline, and those nodes are not connected properly in the service map, despite the trace_ids being propagated to the downstream lambdas.

I grabbed some screen shots.  Function names are again anonymized, though I made sure that the AWS::Lambda and AWS::Lambda::Functions match in colors so it's easier to see what's going on.

Here's a version that's running serially with no goroutines (the first version of the code):

![Serial](https://user-images.githubusercontent.com/1639570/54257884-0ebb4480-451e-11e9-9b59-4cab8bdcb9e0.png)

Here you can see all 4 lambda attempts in the same timeline, exactly as expected.

The following two are from goroutine version of the code (each invocation of lambda.InvokeWithContext in its own goroutine)

Collapsed view, showing that all 4 functions invoked by the warmer appear in the overall trace:
![CollapsedView](https://user-images.githubusercontent.com/1639570/54257125-98b5de00-451b-11e9-807e-38db145dc705.png)

Here's the expanded view of the warmer function's execution from that same trace:
![OneLambda](https://user-images.githubusercontent.com/1639570/54257149-a66b6380-451b-11e9-9384-75873a258303.png)

You can see that only one invocation of Lambda is shows up, rather than the expected 4.   Occasionally a trace will show no invocations, or 2 invocations, but I'd estimate over 90% of them show one.  What I _expect_ to see in this case is all 4 Lambda subnodes similar to the first screenshot, but lined up vertically and showing concurrent execution, rather than the end to end line up that the serial version shows.

**Comment by luluzhao** [comment]
Hello @daroot , we had a bug in our current Go SDK so that you are facing this issue. We will prioritize it and make sure we can address it ASAP.

**Comment by luluzhao** [comment]
Released v1.0.0-rc.12 today which including the fix. Feel free to reopen the issue if you have any questions. Thanks.

**Comment by daroot** [comment]
Confirmed that it fixed the problem for me.  Thanks!
